src/KDSoapClient/KDSoapUdpClient.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2019  Casper Meijn <casper@meijn.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "KDSoapUdpClient.h"
+#include "KDSoapUdpClient_p.h"
+
+#include "KDSoapMessage.h"
+#include "KDSoapMessageReader_p.h"
+#include "KDSoapMessageWriter_p.h"
+#include <QNetworkDatagram>
+#include <QNetworkInterface>
+
+KDSoapUdpClient::KDSoapUdpClient(QObject *parent) : 
+    QObject(parent),
+    d_ptr(new KDSoapUdpClientPrivate(this))
+{
+    Q_D(KDSoapUdpClient);
+    d->socketIPv4 = new QUdpSocket(this);
+    connect(d->socketIPv4, &QUdpSocket::readyRead, [d]() { d->readyReadIPv4(); });
+    d->socketIPv6 = new QUdpSocket(this);
+    connect(d->socketIPv6, &QUdpSocket::readyRead, [d]() { d->readyReadIPv6(); });
+}
+
+bool KDSoapUdpClient::bind(quint16 port) {
+    Q_D(KDSoapUdpClient);
+    bool rc = true;
+    rc = d->socketIPv4->bind(QHostAddress::AnyIPv4, port, QUdpSocket::ShareAddress) && rc;
```

Comment:
Is there a particular reason the ports need to be shared?

Reply:
I was under the impression that other program could use the port as well. This would be useful if multiple programs want to do WS-Discovery.

Reply:
Only for the server though. Clients have arbitrary ports, no? E.g. I see windows10 send Probes from all sorts of ports, but the destination is always 3702.

Reply:
You are right, this appear to work, I will allow both uses then. For some reason my firewall will refuse responses to the random port :-(